### ISSUE TITLE
feat: voice input and pronunciation assessment on flashcards

### ISSUE THREAD

**Issue by huwd** [huwd/learn_hanzi#228]
## Prerequisite

**#222 (multi-axis skill architecture) must be shipped before this issue.** The `skill_masteries` and `skill_review_events` tables are required to record `pronunciation` axis events from voice assessments. Starting implementation without #222 will produce orphaned data.

---

## Problem

Flashcard reviews test recognition but not production. A learner can consistently identify a word's meaning while mispronouncing it — including getting tones wrong — with no feedback signal. Adding voice input to the flashcard closes this gap and feeds data into the `pronunciation` axis from #222.

This is a high-uncertainty feature. The right approach will emerge through iteration. This issue is structured as three phases: prove the interaction model cheaply, add real tone scoring, then improve accuracy.

---

## Why the "STT → wrong characters = wrong tones" hypothesis is unreliable

Modern STT systems are trained to maximise transcription accuracy by applying contextual correction. For isolated single words, a wrong-tone input sometimes produces a different character (妈/马/骂) — but engines like Whisper are trained to infer the most probable character and often normalise tone errors away. This cannot be depended upon as a tone detection mechanism. Phoneme comparison (consonants and vowels) from STT output is useful; tone inference from transcribed characters is not.

---

## Phase 1 — Interaction model + phonetics (no tone scoring)

Prove the mic → record → assess → score interaction on the flashcard for zero or near-zero cost. Tone scoring is deferred; phonetic correctness only.

### Flow

1. Learner taps a mic button on the flashcard back face
2. Browser records a short audio clip
3. Audio is transcribed (see options below)
4. Transcription is converted to pinyin and compared against expected pinyin
5. Score: **correct** (exact phoneme match), **partial** (some phonemes wrong, tones ignored), **wrong** (substantially different)

### Market analysis — STT for phonetic comparison

| Option | Cost | Tone signal | Mandarin quality | Notes |
|---|---|---|---|---|
| **Web Speech API** | Free | None | Good (Chrome sends to Google) | Browser-native; Chrome/Edge only, not Firefox/Safari; audio sent to Google/Microsoft; no tone data returned |
| **Whisper via OpenAI API** | ~$0.006/min | None (text only) | Excellent | Cheapest cloud option; returns text; single words cost fractions of a cent; over-corrects tones |
| **Whisper self-hosted** | Compute only | None | Excellent | MIT licence; tiny/base models fast enough for single words; `whisper.cpp` for efficient server-side inference; no data leaves the app |
| **Whisper.js / transformers.js** | Free | None | Good (smaller model) | Runs entirely in browser via WebAssembly/WebGPU; no API calls; slower but improving rapidly; viable for offline/privacy use |
| **Google Cloud Speech-to-Text** | $0.006/15s | None | Excellent | Word-level confidence scores; no tone assessment |
| **AWS Transcribe** | $0.024/min | None | Good | Higher cost for same capability as Google/Whisper |
| **Vosk** | Free (open source) | None | Moderate | Offline, on-device; Mandarin model available; accuracy below cloud options |
| **AssemblyAI** | $0.012/min | None | Good | No Mandarin-specific features; not the right tool here |

**Recommended for Phase 1**: Web Speech API as the true zero-cost starting point. Self-hosted Whisper as the fallback if browser compatibility is a problem or if the app moves to a privacy-first stance.

---

## Phase 2 — Full pronunciation assessment with tone scoring

Replace or supplement the Phase 1 STT with a purpose-built pronunciation assessment API that returns per-syllable tone scores.

### Market analysis — pronunciation assessment with tone scoring

| Option | Cost | Tone scoring | Mandarin-specific | Notes |
|---|---|---|---|---|
| **Azure Pronunciation Assessment** | Free: 5 hrs/month; then ~$1/hr | Yes — per-syllable tone score (1–4 + neutral) | Yes | Most purpose-built option for learners; returns AccuracyScore, FluencyScore, ToneScore per syllable; JS/REST SDK; straightforward to integrate |
| **SpeechSuper** | Pay-per-use; cheaper than Azure at low volume | Yes | Yes — Chinese company, Mandarin-first | Purpose-built for Chinese pronunciation; returns tone accuracy; less widely documented in English but technically capable |
| **iFlytek (科大讯飞)** | Free tier available | Yes — detailed tone assessment | Yes — industry standard in China | Best raw accuracy for Mandarin; documentation primarily in Chinese; data processed on Chinese servers (sovereignty concern for non-Chinese users); API terms warrant scrutiny |
| **Speechace** | From ~$49/month | Yes | Yes | Phoneme-level scores including tone; priced for EdTech companies, expensive for a personal app |
| **Google Cloud Speech + custom** | $0.006/15s | Indirect (character-level confidence only) | Good | No native tone assessment; would require F0 post-processing to add tone signal |
| **Soapbox Labs** | Custom pricing | No | English-primary | Not suitable |
| **ELSA Speak** | No public API | No | English-primary | Not suitable |

**Recommended for Phase 2**: Azure Pronunciation Assessment. Free tier is generous for a personal app, tone scoring is explicit and per-syllable, and the integration surface is clean.

---

## Phase 3 — F0 pitch analysis for tone detection (no API dependency)

Extract the fundamental frequency (F0) contour from the recorded audio directly and classify the tone shape. This is the DSP approach — no STT, no cloud API, tone inference from acoustic signal alone.

### Market analysis — F0 / pitch analysis

| Option | Cost | Approach | Notes |
|---|---|---|---|
| **Web Audio API + autocorrelation** | Free | In-browser F0 extraction | Pitchfinder.js or custom autocorrelation; classifies contour as flat/rising/dipping/falling; sufficient for clear recordings; struggles with background noise |
| **CREPE.js / TensorFlow.js pitch** | Free | ML-based in-browser F0 | `crepe-js` or `tfjs` port of CREPE; more accurate than autocorrelation; still runs in browser; larger asset to load |
| **Praat (server-side)** | Free | Industry-standard phonetics | CLI tool; accurate F0 extraction; requires a server-side subprocess call; not browser-native |
| **Librosa (Python microservice)** | Free | Server-side F0 | Battle-tested; would require a Python sidecar service; overkill for this use case |
| **WORLD vocoder** | Free | High-quality F0 | C++ library; Ruby FFI possible but complex |
| **Custom tone model (TensorFlow.js)** | Free (training cost) | Train on Mandarin tone examples | Classify tone from MFCC features; requires labelled training data; most accurate long-term but most effort to build |

**Recommended for Phase 3**: CREPE.js for accuracy, Web Audio API autocorrelation for simplicity. Start with autocorrelation; upgrade to CREPE.js if accuracy is insufficient.

---

### Recording

Use the **MediaRecorder API** (browser-native) to capture audio. A minimal Stimulus controller handles mic permission, recording start/stop, and posting the audio blob to the server (or passing to the in-browser STT path).

Do not store audio by default — process and discard. Add an opt-in setting if the user wants recordings retained for later analysis.

### Assessment result storage

```
pronunciation_attempts
  user_id              integer  (FK)
  dictionary_entry_id  integer  (FK)
  assessment_source    string   (web_speech / whisper / azure / speechsuper / f0_local)
  phonetic_score       float    (0.0–1.0)
  tone_score           float    (0.0–1.0, null in Phase 1)
  overall_score        float
  raw_response         jsonb    (full API response for debugging)
  attempted_at         datetime
```

On completion, write a `skill_review_event` for the `pronunciation` axis (#222) using `overall_score`.

### Flashcard integration

Add a mic button to the flashcard back face (alongside the ease rating buttons). Tapping it opens the recording flow. The score is shown briefly before the learner selects their ease rating — they can factor it in or override. Do not auto-submit the card based on the voice score; keep the learner in control.

---

## Relationship to existing work

- **#222 (multi-axis skills)** — hard prerequisite. `pronunciation_attempts` writes to `skill_review_events` for the `pronunciation` axis. This is the primary data source for that dimension.
- **#224 (pronunciation audio)** — stored reference audio could eventually be used as a comparison signal rather than pinyin string matching. Not in scope here but worth noting the asset.
- **#227 (remedial learning)** — persistent low `pronunciation` scores should trigger remedial mode, surfacing tone diagrams and mnemonic aids.
- **#220 (MCP server)** — pronunciation axis weakness can be exposed as an MCP resource for AI-driven practice targeting specific tones.

---

### Phase 1
- [ ] Mic button appears on flashcard back face
- [ ] MediaRecorder captures audio and sends to STT
- [ ] Transcription compared to expected pinyin; correct/partial/wrong score returned
- [ ] `pronunciation_attempts` record created; `skill_review_event` written for `pronunciation` axis
- [ ] Works on Chrome and Safari (iOS mic access confirmed)

### Phase 2
- [ ] Azure Pronunciation Assessment (or chosen alternative) integrated
- [ ] Per-syllable tone score returned and stored in `tone_score`
- [ ] Tone score displayed distinctly from phonetic score on the result
- [ ] Chosen provider and pricing documented in PR

### Phase 3
- [ ] F0 contour extracted in-browser from recorded audio
- [ ] Tone classified (1–4 + neutral) and compared to expected tone
- [ ] Works offline (no API call for tone detection)
- [ ] Accuracy benchmarked against Phase 2 API result on same recordings
